src/data_format_conversions.jl

```diff
@@ -6,32 +6,70 @@
+            if ts_metadata["__metadata__"]["type"] == "DeterministicMetadata" &&
+               !haskey(ts_metadata, "time_series_type")
+                # This will allow deserialization to work.
+                # post_deserialize_conversion will fix the type.
+                ts_metadata["time_series_type"] = Dict(
+                    "__metadata__" => Dict(
+                        "module" => "InfrastructureSystems",
+                        "type" => "AbstractDeterministic",
+                    ),
+                )
+            end
+        end
+    end
+    return
+end
+
+function _convert_data!(raw::Dict{String, Any}, ::Val{Symbol("2.0.0")}, ::Val{Symbol("3.0.0")})
+    for component in raw["data"]["components"]
+        if component["__metadata__"]["type"] == "Bus"
+            component["__metadata__"]["type"] = "ACBus"
+            continue
+        end
+        if component["__metadata__"]["type"] == "HVDCLine"
+            component["__metadata__"]["type"] = "TwoTerminalHVDCLine"
+            continue
+        end
+        if component["__metadata__"]["type"] == "VSCDCLine"
+            component["__metadata__"]["type"] = "TwoTerminalVSCDCLine"
+            continue
+        end
+        if haskey(component, "prime_mover") && haskey(component, "dynamic_injector")
+            component["prime_mover_type"] = pop!(component, "prime_mover")
+        end
+    end
+    return
+end
+
+function _convert_data!(raw::Dict{String, Any}, ::Val{Symbol("1.0.0")}, ::Val{Symbol("3.0.0")})
+    _convert_data!(raw, Val{Symbol("1.0.0")}(), Val{Symbol("2.0.0")}())
+    _convert_data!(raw, Val{Symbol("2.0.0")}(), Val{Symbol("3.0.0")}())
+    return
+end
+
+function _convert_data!(raw::Dict{String, Any}, ::Val{Symbol("1.0.1")}, ::Val{Symbol("3.0.0")})
+    _convert_data!(raw, Val{Symbol("2.0.0")}(), Val{Symbol("3.0.0")}())
+    return
+end
+
 function pre_deserialize_conversion!(raw, sys::System)
     old = raw["data_format_version"]
-    if old == "1.0.0"
-        for component in raw["data"]["components"]
-            for ts_metadata in get(component, "time_series_container", [])
-                if ts_metadata["__metadata__"]["type"] == "DeterministicMetadata" &&
-                   !haskey(ts_metadata, "time_series_type")
-                    # This will allow deserialization to work.
-                    # post_deserialize_conversion will fix the type.
-                    ts_metadata["time_series_type"] = Dict(
-                        "__metadata__" => Dict(
-                            "module" => "InfrastructureSystems",
-                            "type" => "AbstractDeterministic",
-                        ),
-                    )
-                end
-            end
+    if old == DATA_FORMAT_VERSION
+        return
+    else
+        try
+            _convert_data!(raw, Val{Symbol(old)}(), Val{Symbol(DATA_FORMAT_VERSION)}())
+        catch
+            error("conversion of data from $old to $DATA_FORMAT_VERSION is not supported")
```

Comment:
I changed it to return the error in the logs and updates the message.

Reply:
1. The new code has a syntax error. I think you meant `@error err` or `@error “$err”`.
2. This still isn’t ideal. If the exception was caused by a bug in the code, the user will get the string version of the exception message but not the stacktrace.
3. In this scenario we want a very clean error message indicating that this upgrade version path is not supported (which used to be there). Because you changed the conditional logic from an if/else statement to method dispatch, that clarity is lost. A user won’t want to receive a `MethodError` exception. Based on that logic, isn’t the if/else approach better?

Reply:
What we experienced before is that the conversions can fail for multiple reasons. We can do an if-else like this 

```Julia 
            if isa(err, MethodError)
                error(
                "conversion of data from $old to $DATA_FORMAT_VERSION is not supported.",
            )
            else
                @error err
                error(
                    "conversion of data from $old to $DATA_FORMAT_VERSION failed check error log entry.",
                )
            end
end

Reply:
What if there is a different kind of MethodError thrown? The first error message will be very misleading. In the else clause I would call `rethrow()` so that the user gets the exception and backtrace.

Reply:
I mean this function can only have a method error if the versioning is incorrect. We can show the backtrace like this 

```Julia
        catch err
            if isa(err, MethodError)
                error(
                "conversion of data from $old to $DATA_FORMAT_VERSION is not supported.",
            )
            else
                bt = catch_backtrace()
                @error exception = err, bt
                error(
                    "conversion of data from $old to $DATA_FORMAT_VERSION failed check error log entry.",
                )
            end
        end
        @warn(
            "System is saved in the data format version $old will be automatically upgraded to $DATA_FORMAT_VERSION upon saving"
        )
    end
    ```

Reply:
As of _now_ this is the only possible MethodError. New possibilities could be introduced in the future. If someone experiences this misleading error message, they will be very frustrated.

If you want to keep the dispatch approach then I would recommend removing the try/catch entirely. If a failure happens because the old -> new path isn’t implemented, the user will get some `MethodError` exception that indicates a missing implementation of versions. If some other exception happens, the user will receive the normal experience in our codebase when we have a bug: a crash with a backtrace.

Reply:
check [4e1d886](https://github.com/NREL-Sienna/PowerSystems.jl/pull/1012/commits/4e1d886400bccb2d382a93b6d9429cc7f43ad04c)